{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "c8988b10",
   "metadata": {},
   "outputs": [],
   "source": [
    "!pip install gosling\n",
    "import gosling as gos"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8bf62b5e",
   "metadata": {},
   "source": [
    "# Datasets"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "252487de",
   "metadata": {},
   "outputs": [],
   "source": [
    "mv = gos.multivec(\n",
    "    url=\"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\",\n",
    "    row=\"sample\",\n",
    "    column=\"position\",\n",
    "    value=\"peak\",\n",
    "    categories=[\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"],\n",
    "    binSize=2\n",
    ")\n",
    "\n",
    "sv = gos.csv(\n",
    "    url=\"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/rearrangements.bulk.1639.simple.filtered.pub\",\n",
    "    headerNames=[\"chr1\", \"p1s\", \"p1e\", \"chr2\", \"p2s\", \"p2e\", \"type\", \"id\", \"f1\", \"f2\", \"f3\", \"f4\", \"f5\", \"f6\"],\n",
    "    separator=\"\\t\",\n",
    "    genomicFieldsToConvert=[\n",
    "        {\"chromosomeField\": \"chr1\", \"genomicFields\": [\"p1s\", \"p1e\"]},\n",
    "        {\"chromosomeField\": \"chr2\", \"genomicFields\": [\"p2s\", \"p2e\"]}\n",
    "    ]\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81c42e46",
   "metadata": {},
   "source": [
    "# Individual Tracks & Views"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "8e125ba2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-38b398235a684ab1b20f8b7b709d6dd7\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.4/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "        }\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-38b398235a684ab1b20f8b7b709d6dd7');\n",
       "    var spec = {\"alignment\": \"stack\", \"tracks\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"bar\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"brush\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"linkingId\": \"detail-1\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"brush\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"linkingId\": \"detail-2\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"width\": 300, \"height\": 300}, {\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/rearrangements.bulk.1639.simple.filtered.pub\", \"headerNames\": [\"chr1\", \"p1s\", \"p1e\", \"chr2\", \"p2s\", \"p2e\", \"type\", \"id\", \"f1\", \"f2\", \"f3\", \"f4\", \"f5\", \"f6\"], \"separator\": \"\\t\", \"genomicFieldsToConvert\": [{\"chromosomeField\": \"chr1\", \"genomicFields\": [\"p1s\", \"p1e\"]}, {\"chromosomeField\": \"chr2\", \"genomicFields\": [\"p2s\", \"p2e\"]}]}, \"height\": 100, \"mark\": \"withinLink\", \"width\": 300, \"opacity\": {\"value\": 0.6}, \"stroke\": {\"field\": \"chr1\", \"range\": [\"steelblue\"], \"type\": \"nominal\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"field\": \"p1s\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"p2e\", \"type\": \"genomic\"}}], \"layout\": \"circular\", \"spacing\": 1};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  alignment: 'stack',\n",
       "  layout: 'circular',\n",
       "  spacing: 1,\n",
       "  tracks: (View({\n",
       "    alignment: 'overlay',\n",
       "    height: 300,\n",
       "    tracks: (Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "      mark: 'bar',\n",
       "      stroke: StrokeValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 2\n",
       "      }),\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    }), Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "      mark: 'brush',\n",
       "      stroke: StrokeValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 2\n",
       "      }),\n",
       "      x: X({\n",
       "        linkingId: 'detail-1'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    }), Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "      mark: 'brush',\n",
       "      stroke: StrokeValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 2\n",
       "      }),\n",
       "      x: X({\n",
       "        linkingId: 'detail-2'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    })),\n",
       "    width: 300\n",
       "  }), Track({\n",
       "    data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/rearrangements.bulk.1639.simple.filtered.pub', 'headerNames': ['chr1', 'p1s', 'p1e', 'chr2', 'p2s', 'p2e', 'type', 'id', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6'], 'separator': '\\t', 'genomicFieldsToConvert': [{'chromosomeField': 'chr1', 'genomicFields': ['p1s', 'p1e']}, {'chromosomeField': 'chr2', 'genomicFields': ['p2s', 'p2e']}]},\n",
       "    height: 100,\n",
       "    mark: 'withinLink',\n",
       "    opacity: OpacityValue({\n",
       "      value: 0.6\n",
       "    }),\n",
       "    stroke: Stroke({\n",
       "      field: 'chr1',\n",
       "      range: ['steelblue'],\n",
       "      type: 'nominal'\n",
       "    }),\n",
       "    strokeWidth: StrokeWidthValue({\n",
       "      value: 2\n",
       "    }),\n",
       "    width: 300,\n",
       "    x: X({\n",
       "      field: 'p1s',\n",
       "      type: 'genomic'\n",
       "    }),\n",
       "    xe: Xe({\n",
       "      field: 'p2e',\n",
       "      type: 'genomic'\n",
       "    })\n",
       "  }))\n",
       "})"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bar_base = gos.Track(data=mv).mark_bar().encode(\n",
    "    x='start:G',\n",
    "    xe='end:G',\n",
    "    y='peak:Q',\n",
    "    color=gos.Color('sample:N', range=['steelblue', 'salmon', 'steelblue', 'salmon']),\n",
    "    stroke=gos.value('white'),\n",
    "    strokeWidth=gos.value(2),\n",
    ")\n",
    "\n",
    "bar_overview = gos.overlay(\n",
    "    bar_base,\n",
    "    bar_base.mark_brush().encode(x=gos.X(linkingId='detail-1')),\n",
    "    bar_base.mark_brush().encode(x=gos.X(linkingId='detail-2'))\n",
    ").properties(width=300, height=300)\n",
    "\n",
    "link = gos.Track(data=sv).mark_withinLink().encode(\n",
    "    x='p1s:G',\n",
    "    xe='p2e:G',\n",
    "    stroke=gos.Stroke('chr1:N', range=['steelblue']),\n",
    "    strokeWidth=gos.value(2),\n",
    "    opacity=gos.value(0.6)\n",
    ").properties(width=300, height=100)\n",
    "\n",
    "overview = gos.stack(bar_overview, link).properties(layout=\"circular\", spacing=1)\n",
    "\n",
    "overview"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6cb98e9b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-015ab2d7cf6d44f79099ace2a30ca690\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.4/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "        }\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-015ab2d7cf6d44f79099ace2a30ca690');\n",
       "    var spec = {\"arrangement\": \"vertical\", \"views\": [{\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"height\": 100, \"mark\": \"bar\", \"width\": 300, \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"linkingId\": \"detail-1\", \"xDomain\": {\"chromosome\": \"3\"}}, {\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"height\": 100, \"mark\": \"bar\", \"width\": 300, \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"linkingId\": \"detail-2\", \"xDomain\": {\"chromosome\": \"10\"}}], \"spacing\": 20};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  arrangement: 'vertical',\n",
       "  spacing: 20,\n",
       "  views: (View({\n",
       "    linkingId: 'detail-1',\n",
       "    tracks: [Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "      height: 100,\n",
       "      mark: 'bar',\n",
       "      stroke: StrokeValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0\n",
       "      }),\n",
       "      width: 300,\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    })],\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: '3'\n",
       "    })\n",
       "  }), View({\n",
       "    linkingId: 'detail-2',\n",
       "    tracks: [Track({\n",
       "      color: Color({\n",
       "        field: 'sample',\n",
       "        range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "      height: 100,\n",
       "      mark: 'bar',\n",
       "      stroke: StrokeValue({\n",
       "        value: 'white'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 0\n",
       "      }),\n",
       "      width: 300,\n",
       "      x: X({\n",
       "        field: 'start',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'end',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      y: Y({\n",
       "        field: 'peak',\n",
       "        type: 'quantitative'\n",
       "      })\n",
       "    })],\n",
       "    xDomain: GenomicDomain({\n",
       "      chromosome: '10'\n",
       "    })\n",
       "  }))\n",
       "})"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "detail_1 = bar_base.encode(strokeWidth=gos.value(0)).properties(width=300, height=100).view(\n",
    "    linkingId='detail-1',\n",
    "    xDomain=gos.GenomicDomain(chromosome='3'),\n",
    ")\n",
    "\n",
    "detail_2 = bar_base.encode(strokeWidth=gos.value(0)).properties(width=300, height=100).view(\n",
    "    linkingId='detail-2',\n",
    "    xDomain=gos.GenomicDomain(chromosome='10'),\n",
    ")\n",
    "\n",
    "detail_view = gos.vertical(detail_1, detail_2).properties(spacing=20)\n",
    "detail_view"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ff32a2c5",
   "metadata": {},
   "source": [
    "# Compose Views"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "0e1a635e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "<!DOCTYPE html>\n",
       "<html>\n",
       "<head>\n",
       "  <style>.error { color: red; }</style>\n",
       "  <link rel=\"stylesheet\" href=\"https://unpkg.com/higlass@1.11/dist/hglib.css\">\n",
       "</head>\n",
       "<body>\n",
       "  <div id=\"jupyter-gosling-528782c2ae0e40dba78d01fb279af552\"></div>\n",
       "  <script type=\"module\">\n",
       "\n",
       "    async function loadScript(src) {\n",
       "        return new Promise(resolve => {\n",
       "            const script = document.createElement('script');\n",
       "            script.onload = resolve;\n",
       "            script.src = src;\n",
       "            script.async = false;\n",
       "            document.head.appendChild(script);\n",
       "        });\n",
       "    }\n",
       "\n",
       "    async function loadGosling() {\n",
       "        // Manually load scripts from window namespace since requirejs might not be\n",
       "        // available in all browser environments.\n",
       "        // https://github.com/DanielHreben/requirejs-toggle\n",
       "        if (!window.gosling) {\n",
       "            window.__requirejsToggleBackup = {\n",
       "                define: window.define,\n",
       "                require: window.require,\n",
       "                requirejs: window.requirejs,\n",
       "            };\n",
       "            for (const field of Object.keys(window.__requirejsToggleBackup)) {\n",
       "                window[field] = undefined;\n",
       "            }\n",
       "\n",
       "            // load dependencies sequentially\n",
       "            const sources = [\n",
       "                \"https://unpkg.com/react@17/umd/react.production.min.js\",\n",
       "                \"https://unpkg.com/react-dom@17/umd/react-dom.production.min.js\",\n",
       "                \"https://unpkg.com/pixi.js@6/dist/browser/pixi.min.js\",\n",
       "                \"https://unpkg.com/gosling.js@0.9.4/dist/gosling.js\",\n",
       "            ];\n",
       "\n",
       "            for (const src of sources) await loadScript(src);\n",
       "\n",
       "            // restore requirejs after scripts have loaded\n",
       "            Object.assign(window, window.__requirejsToggleBackup);\n",
       "            delete window.__requirejsToggleBackup;\n",
       "        }\n",
       "        return window.gosling;\n",
       "    };\n",
       "\n",
       "    var el = document.getElementById('jupyter-gosling-528782c2ae0e40dba78d01fb279af552');\n",
       "    var spec = {\"arrangement\": \"horizontal\", \"views\": [{\"alignment\": \"stack\", \"tracks\": [{\"alignment\": \"overlay\", \"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"bar\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"brush\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"linkingId\": \"detail-1\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}, {\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"mark\": \"brush\", \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"linkingId\": \"detail-2\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"width\": 300, \"height\": 300}, {\"data\": {\"type\": \"csv\", \"url\": \"https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/rearrangements.bulk.1639.simple.filtered.pub\", \"headerNames\": [\"chr1\", \"p1s\", \"p1e\", \"chr2\", \"p2s\", \"p2e\", \"type\", \"id\", \"f1\", \"f2\", \"f3\", \"f4\", \"f5\", \"f6\"], \"separator\": \"\\t\", \"genomicFieldsToConvert\": [{\"chromosomeField\": \"chr1\", \"genomicFields\": [\"p1s\", \"p1e\"]}, {\"chromosomeField\": \"chr2\", \"genomicFields\": [\"p2s\", \"p2e\"]}]}, \"height\": 100, \"mark\": \"withinLink\", \"width\": 300, \"opacity\": {\"value\": 0.6}, \"stroke\": {\"field\": \"chr1\", \"range\": [\"steelblue\"], \"type\": \"nominal\"}, \"strokeWidth\": {\"value\": 2}, \"x\": {\"field\": \"p1s\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"p2e\", \"type\": \"genomic\"}}], \"layout\": \"circular\", \"spacing\": 1}, {\"arrangement\": \"vertical\", \"views\": [{\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"height\": 100, \"mark\": \"bar\", \"width\": 300, \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"linkingId\": \"detail-1\", \"xDomain\": {\"chromosome\": \"3\"}}, {\"tracks\": [{\"data\": {\"type\": \"multivec\", \"url\": \"https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec\", \"row\": \"sample\", \"column\": \"position\", \"value\": \"peak\", \"categories\": [\"sample 1\", \"sample 2\", \"sample 3\", \"sample 4\"], \"binSize\": 2}, \"height\": 100, \"mark\": \"bar\", \"width\": 300, \"color\": {\"field\": \"sample\", \"range\": [\"steelblue\", \"salmon\", \"steelblue\", \"salmon\"], \"type\": \"nominal\"}, \"stroke\": {\"value\": \"white\"}, \"strokeWidth\": {\"value\": 0}, \"x\": {\"field\": \"start\", \"type\": \"genomic\"}, \"xe\": {\"field\": \"end\", \"type\": \"genomic\"}, \"y\": {\"field\": \"peak\", \"type\": \"quantitative\"}}], \"linkingId\": \"detail-2\", \"xDomain\": {\"chromosome\": \"10\"}}], \"spacing\": 20}], \"title\": \"Overview and Detail Views\"};\n",
       "    var opt = {\"padding\": 0, \"theme\": null};\n",
       "\n",
       "    loadGosling()\n",
       "        .then(gosling => gosling.embed(el, spec, opt))\n",
       "        .catch(err => {\n",
       "            el.innerHTML = `<div class=\"error\">\n",
       "    <p>JavaScript Error: ${error.message}</p>\n",
       "    <p>This usually means there's a typo in your Gosling specification. See the javascript console for the full traceback.</p>\n",
       "</div>`;\n",
       "            throw error;\n",
       "        });\n",
       "  </script>\n",
       "</body>\n",
       "</html>"
      ],
      "text/plain": [
       "View({\n",
       "  arrangement: 'horizontal',\n",
       "  title: 'Overview and Detail Views',\n",
       "  views: (View({\n",
       "    alignment: 'stack',\n",
       "    layout: 'circular',\n",
       "    spacing: 1,\n",
       "    tracks: (View({\n",
       "      alignment: 'overlay',\n",
       "      height: 300,\n",
       "      tracks: (Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "        mark: 'bar',\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 2\n",
       "        }),\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "        mark: 'brush',\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 2\n",
       "        }),\n",
       "        x: X({\n",
       "          linkingId: 'detail-1'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      }), Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "        mark: 'brush',\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 2\n",
       "        }),\n",
       "        x: X({\n",
       "          linkingId: 'detail-2'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      })),\n",
       "      width: 300\n",
       "    }), Track({\n",
       "      data: {'type': 'csv', 'url': 'https://raw.githubusercontent.com/sehilyi/gemini-datasets/master/data/rearrangements.bulk.1639.simple.filtered.pub', 'headerNames': ['chr1', 'p1s', 'p1e', 'chr2', 'p2s', 'p2e', 'type', 'id', 'f1', 'f2', 'f3', 'f4', 'f5', 'f6'], 'separator': '\\t', 'genomicFieldsToConvert': [{'chromosomeField': 'chr1', 'genomicFields': ['p1s', 'p1e']}, {'chromosomeField': 'chr2', 'genomicFields': ['p2s', 'p2e']}]},\n",
       "      height: 100,\n",
       "      mark: 'withinLink',\n",
       "      opacity: OpacityValue({\n",
       "        value: 0.6\n",
       "      }),\n",
       "      stroke: Stroke({\n",
       "        field: 'chr1',\n",
       "        range: ['steelblue'],\n",
       "        type: 'nominal'\n",
       "      }),\n",
       "      strokeWidth: StrokeWidthValue({\n",
       "        value: 2\n",
       "      }),\n",
       "      width: 300,\n",
       "      x: X({\n",
       "        field: 'p1s',\n",
       "        type: 'genomic'\n",
       "      }),\n",
       "      xe: Xe({\n",
       "        field: 'p2e',\n",
       "        type: 'genomic'\n",
       "      })\n",
       "    }))\n",
       "  }), View({\n",
       "    arrangement: 'vertical',\n",
       "    spacing: 20,\n",
       "    views: (View({\n",
       "      linkingId: 'detail-1',\n",
       "      tracks: [Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "        height: 100,\n",
       "        mark: 'bar',\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 0\n",
       "        }),\n",
       "        width: 300,\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      })],\n",
       "      xDomain: GenomicDomain({\n",
       "        chromosome: '3'\n",
       "      })\n",
       "    }), View({\n",
       "      linkingId: 'detail-2',\n",
       "      tracks: [Track({\n",
       "        color: Color({\n",
       "          field: 'sample',\n",
       "          range: ['steelblue', 'salmon', 'steelblue', 'salmon'],\n",
       "          type: 'nominal'\n",
       "        }),\n",
       "        data: {'type': 'multivec', 'url': 'https://server.gosling-lang.org/api/v1/tileset_info/?d=cistrome-multivec', 'row': 'sample', 'column': 'position', 'value': 'peak', 'categories': ['sample 1', 'sample 2', 'sample 3', 'sample 4'], 'binSize': 2},\n",
       "        height: 100,\n",
       "        mark: 'bar',\n",
       "        stroke: StrokeValue({\n",
       "          value: 'white'\n",
       "        }),\n",
       "        strokeWidth: StrokeWidthValue({\n",
       "          value: 0\n",
       "        }),\n",
       "        width: 300,\n",
       "        x: X({\n",
       "          field: 'start',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        xe: Xe({\n",
       "          field: 'end',\n",
       "          type: 'genomic'\n",
       "        }),\n",
       "        y: Y({\n",
       "          field: 'peak',\n",
       "          type: 'quantitative'\n",
       "        })\n",
       "      })],\n",
       "      xDomain: GenomicDomain({\n",
       "        chromosome: '10'\n",
       "      })\n",
       "    }))\n",
       "  }))\n",
       "})"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "gos.horizontal(overview, detail_view).properties(\n",
    "    title='Overview and Detail Views',\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f92a5b77",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
